standards/application/waku-api.md

```diff
@@ -0,0 +1,327 @@
+- Primitive types are `string`, `int`, `bool`, `enum` and `uint`
+- Complex pre-defined types are:
+  - `struct`: object and other nested types.
+  - `array`: iterable object containing values of all the same type.
+  - `result`: an enum type that either contains a value or void (success), or an error (failure); The error is left to the implementor.
+  - `error`: Left to the implementor on whether `error` types are `string` or `object` in the given language.
+- Usage of `result` is RECOMMENDED, usage of exceptions is NOT RECOMMENDED, no matter the language.
+
+### Language mappings
+
+How the API definition should be translated to specific languages.
+
+```yaml
+language_mappings:
+  typescript:
+    naming_convention:
+      - functions: "camelCase"
+      - variables: "camelCase"
+      - types: "PascalCase"
+  nim:
+    naming_convention:
+      - functions: "camelCase"
+      - variables: "camelCase"
+      - types: "PascalCase"
+```
+
+### Application
+
+This API is designed for generic use and ease across all programming languages, for `edge` and `relay` type nodes.
+
+## The Waku API
+
+```yaml
+api_version: "0.0.1"
+library_name: "waku"
+description: "Waku: a private and censorship-resistant message routing library."
+```
+
+### Initialise Waku node
+
+#### Type definitions
+
+```yaml
+types:
+  WakuNode:
+    type: struct
+    description: "A Waku node instance."
+
+  NodeConfig:
+    type: struct
+    fields:
+      mode:
+        type: string
+        constraints: [ "edge", "relay" ]
+        default: "_platform dependent_"
+        description: "The mode of operation of the Waku node. Core protocols used by the node are inferred from this mode."
+      waku_config:
+        type: WakuConfig
+        default: TheWakuNetworkPreset
+      message_confirmation:
+        type: array<string>
+        constraints: [ "store", "filter" ]
+        # Until further dogfooding, assuming default false, usage of SDS should be preferred
+        default: [ "none" ]
+        description: "Whether to apply peer-to-peer reliability strategies to confirm that outgoing message have been received by other peers."
+      networking_config:
+        type: NetworkConfig
+        default: DefaultNetworkingConfig
+      # TODO: to be reviewed, developers have expressed that accepting an object implementing specific traits is useful. 
+      eth_rpc_endpoints:
+        type: array<string>
+        description: "Eth/Web3 RPC endpoint URLs"
+
+  WakuConfig:
+    type: struct
+    fields:
+      seed_nodes:
+        type: array<string>
+        # Default means the node does not bootstrap, e.g. for local development
+        default: []
+        description: "Nodes to connect to; used for discovery bootstrapping and quick connectivity. entree and multiaddr formats are accepted."
+      static_store_nodes:
+        type: array<string>
+        default: []
+        description: "Only the passed nodes are used for store queries, discovered store nodes are discarded."
```

Comment:
sorry if you mentioned it somewhere and I didn't see

question: why discarded? previously we were discussing it in the way that discovered nodes will be de-prioritized and used only if static store nodes are not available

Reply:
I think we can re-adjust at implementation time. Will correct and add todo.